lib/blocs/unread_messages_cubit/unread_messages_cubit.dart

```diff
@@ -0,0 +1,158 @@
+      if (unreadMessages.isEmpty ||
+          unreadThreads.first.lastReply!.createdAt <
+              unreadMessages.first.createdAt) {
+        firstUnreadMsgThread = unreadThreads.first;
+      } else {
+        firstUnreadMsgThread = null;
+      }
+      // if first unread message in thread, emit with firstUnreadThread
+      emit(UnreadMessagesThreadFound(
+          unreadCounter: unreadCounter,
+          unreadThreads: unreadThreads,
+          userLastAccess: userLastAccess,
+          firstUnreadThread: firstUnreadMsgThread));
+    }
+  }
+
+  void listenNotOwnerMessage() {
+    if (state is UnreadMessagesFound) {
+      final currentState = state as UnreadMessagesFound;
+      emit(UnreadMessagesFound(
+          unreadCounter: currentState.unreadCounter + 1,
+          userLastAccess: currentState.userLastAccess));
+    }
+  }
+}
+
+class ThreadUnreadMessagesCubit extends BaseUnreadMessagesCubit {
+  late final _channelUnreadMessagesCubit;
+  // for tracking last access to thread of each unread threads
+  late Map<String, int> userLastAccessThreads;
+
+  ThreadUnreadMessagesCubit(
+      {required ChannelUnreadMessagesCubit channelUnreadMessagesCubit,
+      required ChannelsCubit channelsCubit})
+      : super(channelsCubit: channelsCubit) {
+    _channelUnreadMessagesCubit = channelUnreadMessagesCubit;
+    userLastAccessThreads = Map();
+  }
+
+  void listenOwnerMessage() {
+    super.listenOwnerMessage();
+  }
+
+  void listenNotOwnerMessage() {
+    if (state is UnreadMessagesThreadFound) {
+      final currentState = state as UnreadMessagesThreadFound;
+      emit(UnreadMessagesThreadFound(
+        unreadCounter: currentState.unreadCounter + 1,
+        userLastAccess: currentState.userLastAccess,
+        unreadThreads: currentState.unreadThreads,
+        firstUnreadThread: currentState.firstUnreadThread,
+      ));
+    }
+  }
+
+  void fetchUnreadMessages({required List<Message> messages}) async {
+    if (_channelUnreadMessagesCubit.state is UnreadMessagesThreadFound) {
+      final currentState =
+          _channelUnreadMessagesCubit.state as UnreadMessagesThreadFound;
+      var userLastAccess = currentState.userLastAccess;
+
+      // if this is not first time enter thread
+      final repliedMessage = messages.first;
+      userLastAccess = userLastAccessThreads.putIfAbsent(
+          repliedMessage.id, () => userLastAccess);
+
+      //fetch messsage in thread first
+      int unreadCounter = messages
+          .where((message) => message.createdAt > userLastAccess)
+          .length;
+      if (repliedMessage.createdAt > userLastAccess && unreadCounter != null) {
+        unreadCounter = unreadCounter - 1;
+      }
+
+      if (messages.isNotEmpty) {
+        userLastAccessThreads[repliedMessage.id] = messages.last.createdAt;
+      }
+
+      emit(UnreadMessagesThreadFound(
+        unreadCounter: unreadCounter,
+        userLastAccess: userLastAccess,
```

Comment:
Hm, userLastAccessThreads instead of userLastAccess ?

Reply:
userLastAccess store last time user access to channel(not thread), if i update it it will mark some unread messages in thread to read. For example, i have unread messages in 2 threads, but user click to later thread, so i fetchUnreadMessage in here, userLasAccess get updated, i leave this thread, the older thread will be mark as read now because userLastAccess is updated